flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/stream/StreamingFileCommitter.java

```diff
@@ -0,0 +1,175 @@
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.table.filesystem.TableMetaStoreFactory;
+
+import java.io.Serializable;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeMap;
+
+import static org.apache.flink.table.filesystem.FileSystemOptions.PARTITION_COMMIT_POLICY_SUCCESS_FILE_NAME;
+import static org.apache.flink.table.filesystem.FileSystemOptions.PARTITION_COMMIT_POLICY_TYPE;
+import static org.apache.flink.table.utils.PartitionPathUtils.extractPartitionSpecFromPath;
+import static org.apache.flink.table.utils.PartitionPathUtils.generatePartitionPath;
+
+/**
+ * Committer for {@link StreamingFileWriter}. This is the single (non-parallel) task.
+ */
+public class StreamingFileCommitter extends AbstractStreamOperator<Void>
+		implements OneInputStreamOperator<StreamingFileCommitter.CommitMessage, Void> {
+
+	private static final long serialVersionUID = 1L;
+
+	private final Configuration conf;
+
+	private final List<String> partitionKeys;
+
+	private final TableMetaStoreFactory metaStoreFactory;
+
+	private transient PartitionCommitManager commitManager;
+
+	private transient TaskTracker taskTracker;
+
+	private transient long currentWatermark = Long.MIN_VALUE;
+
+	private transient List<PartitionCommitPolicy> policies;
+
+	public StreamingFileCommitter(
+			List<String> partitionKeys, TableMetaStoreFactory metaStoreFactory, Configuration conf) {
+		this.partitionKeys = partitionKeys;
+		this.metaStoreFactory = metaStoreFactory;
+		this.conf = conf;
+	}
+
+	@Override
+	public void initializeState(StateInitializationContext context) throws Exception {
+		super.initializeState(context);
+		this.commitManager = new PartitionCommitManager(
+				context.isRestored(),
+				context.getOperatorStateStore(),
+				getUserCodeClassloader(),
+				partitionKeys,
+				conf);
+		this.policies = PartitionCommitPolicy.createCommitChain(
+				conf.get(PARTITION_COMMIT_POLICY_TYPE),
+				conf.get(PARTITION_COMMIT_POLICY_SUCCESS_FILE_NAME));
+	}
+
+	@Override
+	public void processElement(StreamRecord<CommitMessage> element) throws Exception {
+		CommitMessage message = element.getValue();
+		for (String partition : message.partitions) {
+			commitManager.addPartition(partition);
+		}
+
+		if (taskTracker == null) {
+			taskTracker = new TaskTracker(message.numberOfTasks);
+		}
+		boolean needCommit = taskTracker.add(message.checkpointId, message.taskId);
+		if (needCommit) {
+			commitPartitions(commitManager.triggerCommit(message.checkpointId));
+		}
+	}
+
+	private void commitPartitions(List<String> partitions) throws Exception {
+		try (TableMetaStoreFactory.TableMetaStore metaStore = metaStoreFactory.createTableMetaStore()) {
```

Comment:
Does this invoke traffic to HMS? If so maybe we should only do it when `partitions` is not empty.